### PR TITLE
Java now uses struct_tag types

### PR DESCRIPTION
**Pull request by kroening**
Note that this does two things (switching base classes to struct_tags and switching the Java front-end to struct_tags). However, they can't be done separately.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 64bfc034386f8bc5c3592adbda7148ade2fb4d9b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87092812
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by tautschnig — APPROVED** [review]
Changes appear to be consistent.

File: src/cpp/cpp_typecheck_bases.cpp [hunk 1]
```diff
@@ -43,18 +43,18 @@ void cpp_typecheckt::typecheck_compound_bases(struct_typet &type)
...
-    if(base_symbol_expr.type().id() == ID_struct_tag)
-      base_symbol_expr.type().id(ID_symbol_type);
+    if(base_symbol_expr.type().id() == ID_symbol_type)
+      base_symbol_expr.type().id(ID_struct_tag);
 
-    if(base_symbol_expr.type().id() != ID_symbol_type)
+    if(base_symbol_expr.type().id() != ID_struct_tag)
     {
       error().source_location=name.source_location();
       error() << "expected type symbol as struct/class base" << eom;
```
Comment (owen-mc-diffblue) [inline src/cpp/cpp_typecheck_bases.cpp:52]: Does this error message need to be updated?

**Comment by smowton** [comment]
Reviewed the first 30% or so, seems ok so far but don't have time to read the whole thing

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 24aa5f90b4dffdadcb1e120aa8a5e477082a3b9a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87616503
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by martin-cs — APPROVED** [review]
The four lines that are in my area seem OK.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: d846585fa7a32eb313c884b297034abdec56a8cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87698036
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by peterschrammel** [comment]
@thk123,  @allredj, please create a TG bump for this.

**Comment by allredj** [comment]
will do

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 5f18b69cbb779cd58ff286d78b2e1887582668f6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87756938
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — CHANGES_REQUESTED** [review]
This breaks test-generator. We're working on a fix but so far we were not able to make test-gen live with that change.

**Review by allredj — COMMENTED** [review]
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: 39e5ab697b0f1b0907e75418feee35666d47f32a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90749965

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4b981568cfab44c796b64f6266215ad50198930d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91761785
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 708a46d338045a6057fc29a5f3d6c3e9cc99934d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92467085
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ddcf2c8e7c82f3c7bc1d6401ef4bdaa12dac94a0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93180215
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request closed without merging**